### PR TITLE
feat(parser): allow parser set event type of handler with middy

### LINKED ISSUE #2407: Feature request: allow parser set event type of handler with middy
### Use case

middy can pass event type like as [this](https://middy.js.org/docs/intro/typescript/). So we can pass inferred event type from zod schema to handler.

### Solution/User Experience

```ts
const detailSchema = z.object({
	date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
	message: z.string(),
});

export const handler = middy()
	// when using parser
	.use(parser({ schema: detailSchema, envelope: EventBridgeEnvelope }))
	// handler will get event type already inferred without type annotation.
	.handler(async (event) => {
		logger.inf(event.detail.message)
	});
```

### Alternative solutions

_No response_

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets)
- [X] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

### PR DESCRIPTION
**Pull request by am29d**
## Summary

This PR introduces type inference for event type of handler when using parser with middy middleware. Based on the combination of provided inputs (schema, envelope, safeParse) the type can change between a single object, an array (i.e. `SqsEnvelope`) and safeParse, where we wrap the result into `ParsedResult`. I have modified the unit tests to use the inference which should have an error, if the type inference breaks.

### Changes

* fixed return types to array for envelopes that return multiple items of provided schema objects
* added `ParseOutput` to narrow down the type of the modified event
* moved `DynamoDBStreamEnvelopeResponse` to types folder
* replaced eslit comments with type casts in unit tests as mentioned in our docs

> Please provide a summary of what's being changed

> Please add the issue number below, if no issue is present the PR might get blocked and not be reviewed

**Issue number:** closes #2407 

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: packages/parser/src/envelopes/sqs.ts [hunk 1]
```diff
@@ -9,15 +9,15 @@ import { ParseError } from '../errors.js';
  *
  *  The record's body parameter is a string, though it can also be a JSON encoded string.
  *  Regardless of its type it'll be parsed into a BaseModel object.
- *
+ *w
```
Comment (dreamorosi) [inline packages/parser/src/envelopes/sqs.ts:12]:
```suggestion
 *
```

File: packages/parser/src/types/parser.ts [hunk 2]
```diff
@@ -34,9 +38,46 @@ type ParsedResult<Input = unknown, Output = unknown> =
...
+> = undefined extends TEnvelope
+  ? ParsedResult<unknown, z.infer<TSchema>>
+  : TEnvelope extends EnvelopeArrayReturnType
+    ? ParsedResult<unknown, z.infer<TSchema>>
+    : ParsedResult<unknown, z.infer<TSchema>[]>;
+
+/**
+ * The output of the parser function, can be either schema inferred type or a ParsedResult
+ */
+type ParseOutput<
```
Comment (dreamorosi) [inline packages/parser/src/types/parser.ts:65]: I wonder if we should call this `ParserOutput` for consistency with the `ParserOptions`.

So far we seem to have called `Parser*` things that related to the parser function and `Parsed*` things that belong to the parsed object.

Wdyt?

**Review by dreamorosi — COMMENTED** [review]
Great work on these types, just left one minor comment about naming (& a suggestion to fix a typo).

Looking forward to merge it!

File: packages/parser/src/types/parser.ts (hunk 2 above)
Comment (am29d) [reply packages/parser/src/types/parser.ts:65]: Makes sense, I will change that. We have also `ParsedResult`, `ParsedResultSuccess` and `ParsedResultError`, should rename them as well 🤔 ?

**Review by dreamorosi — APPROVED** [review]
Thank you for this improvement!

**Pull request merged**